File: TouchNothing/Models/SupabasePayloads.swift
import Foundation

struct AuthSession: Decodable {
    let nickname: String
    let sessionToken: String

    enum CodingKeys: String, CodingKey {
        case nickname
        case sessionToken = "session_token"
    }
}

struct AuthRPCResponse: Decodable {
    let ok: Bool?
    let error: String?
    let nickname: String?
    let sessionToken: String?
    let remainingAttempts: Int?

    enum CodingKeys: String, CodingKey {
        case ok, error, nickname
        case sessionToken = "session_token"
        case remainingAttempts = "remaining_attempts"
    }

    var authSession: AuthSession? {
        guard ok != false,
              let nickname,
              let sessionToken else {
            return nil
        }
        return AuthSession(nickname: nickname, sessionToken: sessionToken)
    }

    var serviceError: SupabaseServiceError? {
        guard ok == false else { return nil }

        if error == "invalid_credentials" {
            return .invalidCredentials(remainingAttempts: remainingAttempts)
        }

        return SupabaseRPCErrorMapper.mapAuthErrorToken(error) ?? .networkFailure
    }
}

struct NicknameParams: Codable {
    let pNickname: String
    let pPin: String

    enum CodingKeys: String, CodingKey {
        case pNickname = "p_nickname"
        case pPin = "p_pin"
    }
}

struct SessionParams: Codable {
    let pNickname: String
    let pDuration: Int
    let pSessionToken: String

    enum CodingKeys: String, CodingKey {
        case pNickname = "p_nickname"
        case pDuration = "p_duration"
        case pSessionToken = "p_session_token"
    }
}

struct UserStatsParams: Codable {
    let pNickname: String

    enum CodingKeys: String, CodingKey {
        case pNickname = "p_nickname"
    }
}

File: app/src/main/java/com/touchnothing/app/data/model/SupabasePayloads.kt
package com.touchnothing.app.data.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class AuthSession(
    val nickname: String,
    @SerialName("session_token") val sessionToken: String,
)

@Serializable
data class AuthRpcResponse(
    val ok: Boolean? = null,
    val error: String? = null,
    val nickname: String? = null,
    @SerialName("session_token") val sessionToken: String? = null,
    @SerialName("remaining_attempts") val remainingAttempts: Int? = null,
)

@Serializable
data class NicknameParams(
    @SerialName("p_nickname") val pNickname: String,
    @SerialName("p_pin") val pPin: String,
)

@Serializable
data class SessionParams(
    @SerialName("p_nickname") val pNickname: String,
    @SerialName("p_duration") val pDuration: Int,
    @SerialName("p_session_token") val pSessionToken: String,
)

@Serializable
data class UserStatsParams(
    @SerialName("p_nickname") val pNickname: String,
)
